File: Java/2024/week10/UnitConverter.java

/*
 Filename: UnitConverter.java
 Author: Mikhail Zubko
 Created: 2024/10/25
 Purpose: 
*/
import java.util.Scanner;

public class UnitConverter {

  // TODO: create conversion methods
  static double CmToInches(double cm, double COEF) {
    return cm / COEF;
  }

  static double InchesToCm(double in, double COEF) {
    return in * COEF;
  }

  static double KmToMiles(double km, double COEF) {
    return km / COEF;
  }

  static double MilesToKm(double mi, double COEF) {
    return mi * COEF;
  }

  // TODO: create a get choice function
  static int Choice(Scanner keyboard) {
    System.out.println("Choose a conversion");
    System.out.println("(1) Centimeters --> Inches");
    System.out.println("(2) Inches --> Centimeters");
    System.out.println("(3) Kilometers --> Miles");
    System.out.println("(4) Miles --> Kilometers");
    int choice;
    while (true) {
      System.out.print("Enter your choice: ");
      try {
        choice = keyboard.nextInt();
      } catch (Exception e) {
        // TODO: handle exception
        System.out.println("Enter an integer");
        keyboard.nextLine();
        continue;
      }
      if (choice < 1 || choice > 4) {
        System.out.println("This option does not exist, try again");
        continue;
      }
      return choice;
    }
  }

  // TODO: create a get source input function
  static double Source(Scanner keyboard, String units) {
    double src;
    while (true) {
      System.out.print("Enter " + units + ": ");
      try {
        src = keyboard.nextDouble();
      } catch (Exception e) {
        // TODO: handle exception
        System.out.println("Enter a float number");
        keyboard.nextLine();
        continue;
      }
      if (src < 0) {
        System.out.println("The input is invalid, enter a positive number");
        continue;
      }
      return src;
    }
  }

  // TODO: main method with menu and user input
  public static void main(String[] args) {
    // TODO: define constants, scanner and variables
    final double IN_CM = 2.54;
    final double MI_KM = 1.60934;
    Scanner keyboard = new Scanner(System.in);
    int choice;
    int another;
    double source = 0;
    double result = 0;
    String src_units = "";
    String dst_units = "";
    // TODO: print a title
    System.out.println("+--------------------------------------+");
    System.out.println("|   Mickey's Multiple Unit Converter   |");
    System.out.println("+--------------------------------------+");
    while (true) {
      choice = Choice(keyboard);
      // TODO: perform calculation and show result
      switch (choice) {
        case 1:
          src_units = "Centimeters";
          dst_units = "Inches";
          source = Source(keyboard, src_units);
          result = CmToInches(source, IN_CM);
          break;
        case 2:
          src_units = "Inches";
          dst_units = "Centimeters";
          source = Source(keyboard, src_units);
          result = InchesToCm(source, IN_CM);
          break;
        case 3:
          src_units = "Kilometers";
          dst_units = "Miles";
          source = Source(keyboard, src_units);
          result = KmToMiles(source, MI_KM);
          break;
        case 4:
          src_units = "Miles";
          dst_units = "Kilometers";
          source = Source(keyboard, src_units);
          result = MilesToKm(source, MI_KM);
      }
      System.out.printf("%.2f %s is %.2f %s\n", source, src_units, result, dst_units);
      // TODO: ask to continue or quit
      System.out.print("Another conversion?\n(1) Yes (any) quit: ");
      try {
        another = keyboard.nextInt();
        if (another == 1) {
          continue;
        }
      } catch (Exception e) {
        // I don't care about this exception here
      }
      System.out.println("Bye!");
      break;
    }
    keyboard.close();
  }
}

File: Python/2024/week7/unit_converter.py
from utils import get_title, get_float
import convert

functions = (
    ("(1) Centimeters --> Inches", convert.cm_to_inches),
    ("(2) Inches --> Centimeters", convert.inches_to_cm),
    ("(3) Kilometers --> Miles", convert.km_to_miles),
    ("(4) Miles --> Kilometers", convert.miles_to_km),
)


def main():
    # TODO: print a heading
    print(get_title("Mickey's Multiple Unit Converter"))
    while True:
        # TODO: display menu
        print("Choose a conversion:")
        for f in functions:
            print(f[0])
        # TODO: get user's choice
        choice = input("Enter your choice: ").strip()
        # TODO: define a target function and the origin unit
        func = None
        origin = None
        match choice:
            case "1":
                origin = "Centimeters"
                func = functions[0][1]
            case "2":
                origin = "Inches"
                func = functions[1][1]
            case "3":
                origin = "Kilometers"
                func = functions[2][1]
            case "4":
                origin = "Miles"
                func = functions[3][1]
            case _:
                print("Wrong input. Try again")
                continue
        # TODO: get user input as a float
        fl_input = get_float(f"Enter {origin}: ")
        if fl_input is None:
            print("You entered nothing.")
        else:
            # TODO: print result
            print(func(fl_input))
        # TODO: ask about another conversion
        ans = input("Another conversion? (y/N): ").lower().strip()
        if ans != "y":
            break


if __name__ == "__main__":
    main()
